src/hooks.ts

```diff
@@ -0,0 +1,141 @@
+  overrideUserId?: string,
+  overrideAttributes?: optimizely.UserAttributes,
+}
+
+interface UseFeature {
+  (
+    featureKey: string,
+    options?: UseFeatureOptions,
+    overrides?: UseFeatureOverrides,
+  ): [
+    UseFeatureState["isEnabled"],
+    UseFeatureState["variables"],
+    ClientReady,
+    DidTimeout
+  ]
+}
+
+/**
+ * A React Hook that retrieves the status of a feature flag and its variables, optionally
+ * auto updating those values based on underlying user or datafile changes.
+ * 
+ * Note: The react client can become ready AFTER the timeout period. 
+ *       ClientReady and DidTimeout provide signals to handle this scenario.
+ */
+export const useFeature : UseFeature = (featureKey, options = {}, overrides = {}) => {
+  const { isServerSide, optimizely, timeout } = useContext(OptimizelyContext);
+  if (!optimizely) {
+    throw new Error('optimizely prop must be supplied via a parent <OptimizelyProvider>');
+  }
+  const finalReadyTimeout: number | undefined =
+      options.timeout !== undefined ? options.timeout : timeout;
+
+  // Helper function to return the current values for isEnabled and variables.
+  const getCurrentValues = useCallback(() => ({
+    isEnabled: optimizely.isFeatureEnabled(featureKey, overrides.overrideUserId, overrides.overrideAttributes),
+    variables: optimizely.getFeatureVariables(featureKey, overrides.overrideUserId, overrides.overrideAttributes),
+  }), [featureKey, overrides]);
+
+  // Set the initial state immediately serverSide
+  const [ data, setData ] = useState<UseFeatureState>(() => {
+    if (isServerSide) {
+      return getCurrentValues();
+    }
+    return { isEnabled: false, variables: {}};
+  });
+
+  const [clientReady, setClientReady] = useState(isServerSide ? true : false);
+  const [didTimeout, setDidTimeout] = useState(false);
+
+  useEffect(() => {
+    const cleanupFns: Array<() => void> = [];
+
+    optimizely.onReady({ timeout: finalReadyTimeout }).then((res: OnReadyResult) => {
+      if (res.success) {
+        // didTimeout=false
+        useFeatureLogger.info(`feature="${featureKey}" successfully set for user="${optimizely.user.id}"`);
+        return;
+      }
+      setDidTimeout(true);
+      useFeatureLogger.info(
+        `feature="${featureKey}" could not be set before timeout of ${finalReadyTimeout}ms, reason="${res.reason || ''}"`,
+      );
+      // Since we timed out, wait for the dataReadyPromise to resolve before setting up.
+      return res.dataReadyPromise!.then(
+        () => {
+          useFeatureLogger.info(
+            `feature="${featureKey}" is now set, but after timeout.`,
+          );
+        });
+    })
+    .then(() => {
+      setClientReady(true);
+      setData(getCurrentValues());
+      if (options.autoUpdate) {
+        cleanupFns.push(
+          setupAutoUpdateListeners(optimizely, 'feature', featureKey, useFeatureLogger, () => {
+            if (cleanupFns.length) {
+              setData(getCurrentValues());
```

Comment:
Will this be passed the `autoUpdate` option more than once?

Reply:
Not sure I understand your question here. Can you elaborate?

Reply:
It looked like `setData` was called on initial update but I wasn't sure if the `autoUpdate` needed to be passed again. Upon taking another look I see that `setupAutoUpdateListeners` is just called anytime the user / instance is updated so nevermind.